chainer/testing/function_link.py

```diff
@@ -0,0 +1,1116 @@
+
+    The concrete class can override the following attributes to control the
+    behavior of the tests.
+
+    ``skip_forward_test`` (bool):
+        Whether to skip forward computation test. ``False`` by default.
+
+    ``skip_backward_test`` (bool):
+        Whether to skip backward computation test. ``False`` by default.
+
+    ``skip_double_backward_test`` (bool):
+        Whether to skip double-backward computation test. ``False`` by default.
+
+    ``dodge_nondifferentiable`` (bool):
+        Enable non-differentiable point detection in numerical gradient
+        calculation. If the inputs returned by ``generate_inputs`` turns
+        out to be a non-differentiable point, the test will repeatedly resample
+        inputs until a differentiable point will be finally sampled.
+        ``False`` by default.
+
+    ``contiguous`` (None or 'C'):
+        Specifies the contiguousness of incoming arrays (i.e. inputs, output
+        gradients, and the second order input gradients). If ``None``, the
+        arrays will be non-contiguous as long as possible. If ``'C'``, the
+        arrays will be C-contiguous. ``None`` by default.
+
+    .. note::
+
+       This class assumes :func:`chainer.testing.inject_backend_tests`
+       is used together. See the example below.
+
+    .. admonition:: Example
+
+       .. testcode::
+
+          @chainer.testing.inject_backend_tests(
+              None,
+              [
+                  {},  # CPU
+                  {'use_cuda': True},  # GPU
+              ])
+          class TestReLU(chainer.testing.FunctionTestCase):
+
+              # ReLU function has a non-differentiable point around zero, so
+              # dodge_nondifferentiable should be set to True.
+              dodge_nondifferentiable = True
+
+              def generate_inputs(self):
+                  x = numpy.random.uniform(-1, 1, (2, 3)).astype(numpy.float32)
+                  return x,
+
+              def forward(self, inputs, device):
+                  x, = inputs
+                  return F.relu(x),
+
+              def forward_expected(self, inputs):
+                  x, = inputs
+                  expected = x.copy()
+                  expected[expected < 0] = 0
+                  return expected,
+
+    .. seealso:: :class:`~chainer.testing.LinkTestCase`
+
+    """
+
+    def test_forward(self, backend_config):
+        """Tests forward computation."""
+        self.run_test_forward(backend_config)
+
+    def test_backward(self, backend_config):
+        """Tests backward computation."""
+        self.run_test_backward(backend_config)
+
+    def test_double_backward(self, backend_config):
+        """Tests double-backward computation."""
+        self.run_test_double_backward(backend_config)
+
+
+class _LinkTestImpl(object):
+
+    contiguous = None
+    backend_config = None
+
+    # List of parameter names represented as strings.
+    # I.e. ['gamma', 'beta'] for BatchNormalization.
+    param_names = []
```

Comment:
It's better to use tuple. It should be immutable.

Reply:
Thanks, fixed.